### DISCUSSION TITLE
Assign properties to refined mesh elements

### DISCUSSION THREAD

**Discussion by ritamcj** [idaholab/moose#18837] [Q&A General] [answered]
Hi everyone,
I am attempting to model nucleation in a multiphase system through a crystal plasticity-phase field model. I looked at [refine.txt](https://github.com/idaholab/moose/files/7175897/refine.txt) in which nucleation triggers concentration changes based on the concentration gradient at the element interfaces (GradientJumpIndicator) and also, mesh refinement based on ValueThresholdMarker. For my CP based model, I am inputting Euler angles based on Element ID through ElementPropertyReadFile. Initially, I have 4 elements with unique orientations. **With mesh refinement, I need to assign random orientations for the new elements.** I kept extra orientations (~5000) in my orientation input file for the same. However, I see the attached error as soon as the mesh is refined due to nucleation. I printed the element ID 'jelem' and it goes to values as high as 160. I think the no. of elements "nelem" needs to be updated. Kindly suggest strategies to tackle this problem. I found this discussion somewhat relevant https://github.com/idaholab/moose/issues/3080.
Warm regards,
Ritam

![Screenshot from 2021-09-16 13-29-29](https://user-images.githubusercontent.com/81842276/133574956-94d70a16-b264-424b-a460-bbd06d9df0dd.png)
![Screenshot from 2021-09-16 13-53-39](https://user-images.githubusercontent.com/81842276/133577562-622edc87-224e-42d5-bcd2-b037b154c3bc.png)

**Comment by ngrilli** [comment, marked as answer]
Dear @ritamcj 

Yes, you are right the line:
_nelem = _mesh.nElem();

is fixing the number of elements in the constructor of ElementPropertyReadFile, therefore you get the error in:
ElementPropertyReadFile::getElementData

Just an idea but not sure if that would work,
what happens if you update the number of elements inside ElementPropertyReadFile::getElementData

Meaning you define another class and you add the line

_nelem = _mesh.nElem();

at the beginning of getElementData

In that case the number of elements may be updated correctly, but I am not sure, just try.
I think the problem is that in the constructor the number of elements is assigned only once in the simulation.

Nicolò

**Reply by ritamcj** [reply]
Hi Nicolo,
Thank you very much for your detailed response. It did work :) I used a dummy variable to read the updated mesh elements as:
```
Real
ElementPropertyReadFile::getElementData(const Elem * elem, unsigned int prop_num) const
{
  unsigned int num_elem = _mesh.nElem();   
  
  unsigned int jelem = elem->id();

  if (jelem >= num_elem)
    mooseError("Element ID", jelem, " greater than than total number of element in mesh ", num_elem);
  return _reader.getData(jelem)[prop_num];
}
```
Warm regards,
Ritam
